### PR TITLE
Some Optimizations

### PR DESCRIPTION
**Pull request by adi-kmt**

### CODE REVIEW

**Comment by IsNoobgrammer** [comment]
Hi @adi-kmt 

I have a update to fix some bugs and normalize the tokens ; 
so i think your commit may not be directly mergeable after that ;

Could you please create a new PR after latest patch ; which should be done by 00:00 IST

File: src/heuristics.py [hunk 1]
```diff
@@ -10,185 +10,216 @@
...
+        # Move results to CPU
+        return global_embedding_input.cpu(), global_embedding_output.cpu() if global_embedding_output is not None else None
 
     except Exception as e:
-        print(f"Warning: Error during FAISS search/processing for '{query_token_str}': {e}")
+        print(f"Error calculating global embedding for token '{query_token_str}': {e}")
         return None, None
+    
 
+def precompute_subtoken_embeddings_array(
```
Comment (IsNoobgrammer) [inline src/heuristics.py:107]: what is the need of this function ? 
does it provide any optimization ;

Also as a sub-note ; can we keep changes to minimum like only update what's needed

File: src/heuristics.py [hunk 2]
```diff
@@ -10,185 +10,216 @@
...
 import faiss
-from typing import Optional,Tuple
+from typing import List, Optional,Tuple
+import numpy as np
+import torch
+import torch.nn.functional as F
+import faiss
+from typing import Optional, Tuple, Dict, Set
 
 def calculate_global_embedding(
```
Comment (IsNoobgrammer) [inline src/heuristics.py:20]: I have a good idea can you add PR ; after patch so it allows user to select either IndexFlatIp (accuracy) or indexhnswFlat (speed)

Comment (adi-kmt) [reply src/heuristics.py:20]: haan will do that

File: src/heuristics.py (hunk 1 above)
Comment (adi-kmt) [reply src/heuristics.py:107]: You can basically do this 
`subtoken_embeddings_array, subtoken_valid_mask = precompute_subtoken_embeddings_array(
        old_tokenizer, subtoken_embeds_cache, embedding_dim, max_id
    )
    `
    Then pass it into
    `
    calculate_local_embedding_optimized
    `
    But this cache thing can be removed
